modules/ding_campaign_plus/js/ding_campaign_plus.js

```diff
@@ -24,6 +24,8 @@
           if (response) {
             campaign.replaceWith(response);
             wrapper.slideDown('fast');
+            // We add new content so attach behaviors.
+            Drupal.attachBehaviors($('<div></div>').html(response));
```

Comment:
Attaching behaviors to something that is not bound to the document seems odd to me although I am not sure about how `replaceWith()` handles the response by reference or by value.

How about attaching behaviors to the wrapper instead?

Reply:
Had some trouble getting this to work otherwise. Using the wrapper might work. will try it out!

Reply:
The wrapper has HTML which is not in response and is already in the DOM, so I opted for finding the nearest parent() of the campaign we just inserted and use that in attachBehaviors.

Went down bit of a rabbit hole with this. If you take a look at how Drupal core attaches behaviors when new content is added with ajax commands in insert. From Drupal core /includes/ajax.js (around lines 514-534):

```
// We don't know what response.data contains: it might be a string of text
// without HTML, so don't rely on jQuery correctly iterpreting
// $(response.data) as new HTML rather than a CSS selector. Also, if
// response.data contains top-level text nodes, they get lost with either
// $(response.data) or $('<div></div>').replaceWith(response.data).
var new_content_wrapped = $('<div></div>').html(response.data);
var new_content = new_content_wrapped.contents();

// For legacy reasons, the effects processing code assumes that new_content
// consists of a single top-level element. Also, it has not been
// sufficiently tested whether attachBehaviors() can be successfully called
// with a context object that includes top-level text nodes. However, to
// give developers full control of the HTML appearing in the page, and to
// enable Ajax content to be inserted in places where DIV elements are not
// allowed (e.g., within TABLE, TR, and SPAN parents), we check if the new
// content satisfies the requirement of a single top-level element, and
// only use the container DIV created above when it doesn't. For more
// information, please see http://drupal.org/node/736066.
if (new_content.length != 1 || new_content.get(0).nodeType != 1) {
  new_content = new_content_wrapped;
}
```

So it looks like it's also using the '<div></div>'-wrapper in certain cases, but it appears to not be because of the same reasons that I'm experiencing.

I just can't understand how it works when it's not using '<div></div>'-wrapper. The find method does ignore the root-element: http://danielnouri.org/notes/2011/03/14/a-jquery-find-that-also-finds-the-root-element/

Anyways, http://drupal.org/node/736066 is a rather interesting discussion about how to attach behaviors when new content is added.

Reply:
:+1: Thanks for the elaboration 😅